osu.Game/Graphics/UserInterface/SegmentedGraph.cs

```diff
@@ -0,0 +1,320 @@
+            }
+        }
+
+        public readonly Colour4[] TierColours;
+
+        private Texture texture = null!;
+        private IShader shader = null!;
+
+        [BackgroundDependencyLoader]
+        private void load(IRenderer renderer, ShaderManager shaders)
+        {
+            texture = renderer.WhitePixel;
+            shader = shaders.Load(VertexShaderDescriptor.TEXTURE_2, FragmentShaderDescriptor.TEXTURE);
+        }
+
+        protected override void Update()
+        {
+            base.Update();
+
+            if (graphNeedsUpdate)
+            {
+                recalculateSegments();
+                Invalidate(Invalidation.DrawNode);
+            }
+        }
+
+        private void recalculateTiers(T[]? arr)
+        {
+            if (arr == null || arr.Length == 0)
+            {
+                tiers = Array.Empty<int>();
+                return;
+            }
+
+            float[] floatValues = arr.Select(v => Convert.ToSingle(v)).ToArray();
+
+            // Shift values to eliminate negative ones
+            float min = floatValues.Min();
+
+            if (min < 0)
+            {
+                for (int i = 0; i < floatValues.Length; i++)
+                    floatValues[i] += Math.Abs(min);
+            }
+
+            // Normalize values
+            float max = floatValues.Max();
+
+            for (int i = 0; i < floatValues.Length; i++)
+                floatValues[i] /= max;
+
+            // Deduce tiers from values
+            tiers = floatValues.Select(v => (int)Math.Floor(v * tierCount)).ToArray();
+        }
+
+        private void recalculateSegments()
+        {
+            segments.Clear();
+
+            if (tiers.Length == 0)
+            {
+                segments.Add(0, 0, 1);
+                return;
+            }
+
+            for (int i = 0; i < tiers.Length; i++)
+            {
+                for (int tier = 0; tier < tierCount; tier++)
+                {
+                    if (tier < 0)
+                        continue;
+
+                    // One tier covers itself and all tiers above it.
+                    // By layering multiple transparent boxes, higher tiers will be brighter.
+                    // If using opaque colors, higher tiers will be on front, covering lower tiers.
+                    if (tiers[i] >= tier)
+                    {
+                        if (!segments.IsTierStarted(tier))
+                            segments.StartSegment(tier, i * 1f / tiers.Length);
+                    }
+                    else
+                    {
+                        if (segments.IsTierStarted(tier))
+                            segments.EndSegment(tier, i * 1f / tiers.Length);
+                    }
+                }
+            }
+
+            segments.EndAllPendingSegments();
+            segments.Sort();
+        }
+
+        private Colour4 tierToColour(int tier) => tier >= 0 ? TierColours[tier] : new Colour4(0, 0, 0, 0);
```

Comment:
Nit but I'd rename this to ```getColourForTier()``` fits better with the general naming in the project and is more intuitive IMO

Reply:
agree